src/Service/EolValidator.php

```diff
@@ -0,0 +1,218 @@
+        ServiceFactory $serviceFactory,
+        ElasticSearch $elasticSearch
+    ) {
+        $this->fileList = $fileList;
+        $this->serviceFactory = $serviceFactory;
+        $this->elasticSearch = $elasticSearch;
+    }
+
+    /**
+     * Validate the EOL of a given service and version by error level.
+     *
+     * @param int $errorLevel
+     * @return array
+     * @throws ServiceMismatchException
+     */
+    public function validateServiceEol(int $errorLevel): array
+    {
+        $errors = [];
+
+        foreach ($this->services as $serviceName) {
+            $serviceVersion = $this->getServiceVersion($serviceName);
+
+            // Validate EOL for each service.
+            if ($validationResult = $this->validateService($serviceName, $serviceVersion, $errorLevel)) {
+                $errors[] = $validationResult;
+            }
+        }
+
+        return $errors;
+    }
+
+    /**
+     * Validates a given service and version.
+     *
+     * @param string $serviceName
+     * @param string $serviceVersion
+     * @param int $errorLevel
+     * @return string
+     */
+    public function validateService(string $serviceName, string $serviceVersion, int $errorLevel) : string
+    {
+        // Get the EOL configurations for the current service.
+        $serviceConfigs = $this->getServiceEolConfigs($serviceName);
+
+        // Check if configurations exist for the current service and version.
+        $versionConfigs = array_filter($serviceConfigs, function ($v) use ($serviceVersion) {
+            return Semver::satisfies($serviceVersion, sprintf('%s.x', $v['version']));
+        });
+
+        // If there are no configurations found for the current service and version,
+        // or if an EOL is not defined or is invalid, return a message with details.
+        if (!$versionConfigs || $versionConfigs[array_key_first($versionConfigs)]['eol'] === null) {
+            return $errorLevel === ValidatorInterface::LEVEL_WARNING ? sprintf(
+                'Unknown or invalid EOL defined for %s %s',
+                $serviceName,
+                $serviceVersion
+            ) : '';
+        }
+
+        $eolDate = Carbon::createFromTimestamp($versionConfigs[array_key_first($versionConfigs)]['eol']);
+
+        // If the EOL is in the past, issue a warning.
+        // If the EOL is in the future, but within a three month period, issue a notice.
+        return $this->getServiceEolNotifications($eolDate, $errorLevel, $serviceName, $serviceVersion);
+    }
+
+    /**
+     * Gets the version of a given service.
+     *
+     * @param string $serviceName
+     * @return string
+     * @throws ServiceMismatchException
+     */
+    private function getServiceVersion(string $serviceName) : string
+    {
+        switch ($serviceName) {
+            case ServiceInterface::NAME_PHP:
+                $serviceVersion = PHP_VERSION;
+                break;
+            case ServiceInterface::NAME_ELASTICSEARCH:
```

Comment:
Service factory can create  `elasticsearch` service, why do you need a custom scenario?